design/baremetal-operator/configmap_for_host_inspection_data.md

```diff
@@ -0,0 +1,148 @@
+
+### Implementation Details/Notes/Constraints
+
+The life of a ConfigMap is tightly coupled to the corresponding BareMetalHost.
+Creating or deleting a BareMetalHost will result in creation or deletion of a
+ConfigMap respectively.
+
+**Note:** [cross-namespace owner references are not allowed](https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/#owner-references-in-object-specifications). Thus, ConfigMap and BareMetalHost
+must be in the same namespace.
+
+- BareMetalHost is created
+- Operator creates a ConfigMap with the same name and namespace as BareMetalHost
+    and sets ownerReferences
+- Operator writes name of the ConfigMap to `spec.hardwareInfo` of the BareMetalHost
+- Provisioning requested, operator fills out the ConfigMap with the host
+    inspection data
+- Deprovisioning requested, operator cleans up the inspection data from the
+    ConfigMap
+- BareMetalHost deletion requested, operator deletes the ConfigMap and BareMetalHost
+
+#### BareMetalHost API changes
+
+- `status.hardware` field should be removed ⚠️
+- `spec.hardwareInfo` field should be added, to keep the name of ConfigMap as a
+    reference
+
+### Risks and Mitigations
+
+If an unintended user gains an access to the cluster and deletes the ConfigMap,
+BMO will notice the absence of the ConfigMap in the next reconciliation loop,
+and will try to recreate it. This will potentially trigger another provisioning.
+
+### Work Items
+
+- Remove `hardware` field from the Status of BareMetalHost API
+- Add `hardwareInfo` to the Spec of BareMetalHost API
+- Modify the operator code
+- Clean up statusAnnotation related code from BMO and CAPM3
+- Add unit tests
+- Update documentation
+- Update dependencies in Cluster-api-provider-metal3, metal3-dev-env GitHub
+    repositories
+- Give a notice to the community in advance about the change, because some
+    external code bases might have tight dependencies
+
+### Dependencies
+
+None
+
+### Test Plan
+
+Unit tests should be added for new functions and integration test scripts
+should be updated.
+
+### Upgrade / Downgrade Strategy
+
+From a user perspective, this feature doesn't introduce any breaking changes
+to existing deployments, except when there is reliance on the `status.hardware`
+field from external code base.
+
+### Version Skew Strategy
+
+I can not foresee any version skew issues.
+
+## Drawbacks
+
+None. Is there anything you can imagine?
+
+## Alternatives
+
+- Use Secret instead of ConfigMap. However, inspection data doesn't contain
+    confidential information, so there is no benefit we would get by having
+    the data encoded.
+- Create a new Custom Resource type, and write the data into its spec. Since
+    there is no control loop specifically on the inspection data and we only
+    write, read or delete it, it seems redundant to introduce a new CRD for
+    this purpose when Kubernetes offers built in objects like ConfigMap/Secret.
```

Comment:
I disagree here. The value of a CRD is not that you can write a controller to reconcile it (you can write a controller to reconcile anything, including ConfigMaps), it's that the schema is documented, verified, and versioned, and that specific RBAC rules can be applied to it.

I think defining a new CRD that puts the hardware details in the Spec instead of the Status is the correct solution. (This also solves the name collision issue, which also makes it easier to roll out in v1alpha1.)

I documented how I think this should work in [the API decomposition proposal](https://docs.google.com/document/d/1FL8oA0_WNcPdiC-0zrToinl8aol-NPQUjWe-OkcvTqA/edit#heading=h.vb3zlo2h9t7).

Reply:
@zaneb thanks a lot for the quick review. I agree now that it would make more sense to have a new CRD instead of ConfigMap, which was something I didn't have in my mind while designing. 
If we go with a CRD and keep the hardwareDetails for some time (not to break current users and give enough time for others to prepare), is there some other blocker you can see? I can re-write the proposal, and request another review and after addressing comments, can we right away start implementation or do we have some major plans with regard to API change that this work should hold on?

Reply:
I don't think there's a blocker. We have added a bunch of new CRDs without bumping the version. Removing the existing HardwareDetails would be a blocker, but I think the answer is just to copy them from the CRD if it is available into the Status, and vice-versa, but exclude it from the status annotation. That way it survives the pivot (via the CRD), doesn't break the status annotation size limit, and presents the same behaviour for existing API users.

Reply:
Awesome. Thanks again, I will ping once I updated the proposal.

Reply:
Updated accordingly.